src/resolver.rs

```diff
@@ -188,14 +181,28 @@ pub struct CriteriaFailureSet {
     pub all: CriteriaSet,
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
+pub enum CriteriaNamespace {
+    Local,
+    Foreign(ImportName),
+}
+
+#[derive(Debug, Clone)]
+pub struct CriteriaInfo {
+    pub namespace: CriteriaNamespace,
+    pub raw_name: CriteriaName,
```

Comment:
It looks like this is only used when building up the mapper and isn't accessed after that, it might be nice to not hold onto the raw name if we don't need it.

The `namespace` is also used when handling cross-namespace implies cycles, though if we dedupe at the index level when doing mappings, we could also probably remove that field, as it's not accessed other than there and when building up the mapper (which I believe could be done in a single pass).

Reply:
it's also used to do the latebound deduping when getting a list of set criteria, but yeah if we did early-bound deduping we might not need it..?